plugins/cros-ec/fu-cros-ec-firmware.c

```diff
@@ -0,0 +1,140 @@
+
+	self->sections[0].name = "RO";
+	self->sections[1].name = "RW";
+
+	if (!fu_firmware_parse (fmap_firmware, fw, flags, error))
+		return FALSE;
+
+	for (gsize i = 0; i < G_N_ELEMENTS (self->sections); i++) {
+		gboolean rw = FALSE;
+		FuCrosEcFirmwareSection *section = &self->sections[i];
+		const gchar *fmap_name;
+		const gchar *fmap_fwid_name;
+		g_autoptr(FuFirmwareImage) img = NULL;
+		g_autoptr(FuFirmwareImage) fwid_img = NULL;
+		g_autoptr(GBytes) payload_bytes = NULL;
+		g_autoptr(GBytes) fwid_bytes = NULL;
+
+		if (g_strcmp0 (section->name, "RO") == 0) {
+			fmap_name = "EC_RO";
+			fmap_fwid_name = "RO_FRID";
+		} else if (g_strcmp0 (section->name, "RW") == 0) {
+			rw = TRUE;
+			fmap_name = "EC_RW";
+			fmap_fwid_name = "RW_FWID";
+		} else {
+			g_set_error_literal (error,
+					     G_IO_ERROR,
+					     G_IO_ERROR_INVALID_DATA,
+					     "incorrect section name");
+			return FALSE;
+		}
+
+		img = fu_firmware_get_image_by_id (fmap_firmware,
+						   fmap_name, error);
+		if (img == NULL) {
+			g_prefix_error (error, "%s image not found: ",
+					fmap_name);
+			return FALSE;
+		}
+
+		fwid_img = fu_firmware_get_image_by_id (fmap_firmware,
+							fmap_fwid_name, error);
+		if (fwid_img == NULL) {
+			g_prefix_error (error, "%s image not found: ",
+					fmap_fwid_name);
+			return FALSE;
+		}
+		fwid_bytes = fu_firmware_image_write (fwid_img, error);
+		if (fwid_bytes == NULL) {
+			g_prefix_error (error,
+					"unable to get bytes from %s: ",
+					fmap_fwid_name);
+			return FALSE;
+		}
+		if (!fu_memcpy_safe ((guint8 *) section->version,
+				     FMAP_STRLEN, 0x0,
+				     g_bytes_get_data (fwid_bytes, NULL),
+				     g_bytes_get_size (fwid_bytes), 0x0,
+				     g_bytes_get_size (fwid_bytes), error))
+			return FALSE;
+
+		payload_bytes = fu_firmware_image_write (img, error);
+		if (payload_bytes == NULL) {
+			g_prefix_error (error,
+					"unable to get bytes from %s: ",
+					fmap_name);
+			return FALSE;
+		}
+		section->offset = fu_firmware_image_get_addr (img);
+		section->size = g_bytes_get_size (payload_bytes);
+		fu_firmware_image_set_version (img, section->version);
+		fu_firmware_add_image (firmware, img);
+
+		if (rw) {
+			if (!fu_cros_ec_parse_version (section->version,
+						       &self->version,
+						       error)) {
+				g_prefix_error (error,
+						"failed parsing firmware's version: %32s: ",
+						section->version);
+				return FALSE;
+			}
+			fu_firmware_set_version (firmware,
+						 self->version.triplet);
+		}
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static void
+fu_cros_ec_firmware_init (FuCrosEcFirmware *self)
+{
+}
+
+static void
+fu_cros_ec_firmware_class_init (FuCrosEcFirmwareClass *klass)
+{
+	FuFirmwareClass *klass_firmware = FU_FIRMWARE_CLASS (klass);
+	klass_firmware->parse = fu_cros_ec_firmware_parse;
+}
+
+FuFirmware *
+fu_cros_ec_firmware_new (void)
+{
+	return FU_FIRMWARE (g_object_new (FU_TYPE_CROS_EC_FIRMWARE, NULL));
```

Comment:
do you actually need the `FU_FIRMWARE (` macro?  I think just `g_object_new` is sufficient

Reply:
I don't think it's required, but I also don't have a problem with it being there. If `FU_TYPE_CROS_EC_FIRMWARE` isn't squeezable down into a FuFirmware we have bigger problems. :)

Reply:
Done. Removed.